cronjob_scripts/trigger-package-build.py

```diff
@@ -0,0 +1,182 @@
+            )
+
+        workflow_run_input = {
+            "crate": crate,
+            "version": version["vers"],
+            "target_arch": target_arch,
+            "features": features,
+            "no_default_features": no_default_features,
+            "build_os": build_os,
+            "branch": branch,
+        }
+        print(f"Attempting to build {crate} {version['vers']} for {target_arch}")
+        subprocess.run(
+            ["gh", "workflow", "run", "build-package.yml", "--json", f"--ref={branch}"],
+            input=json.dumps(workflow_run_input).encode(),
+            check=True,
+        )
+        builds_scheduled += 1
+        if builds_scheduled >= check_limit:
+            print("Scheduled enough builds, exiting")
+            break
+        time.sleep(30)
+
+
+def get_popular_crates():
+    """
+    List of crates that are popular and should be checked in addition to the requested crates.
+
+    In the future, this list might be generated from the crates.io db dump (see
+    https://github.com/cargo-bins/cargo-quickinstall/issues/268#issuecomment-2329308074),
+    but probably not every time we run this script. We will need to think about this.
+    """
+    with open(
+        Path(__file__).resolve().parent.parent.joinpath("popular-crates.txt"), "r"
+    ) as file:
+        return [
+            stripped
+            for line in file
+            if (stripped := line.strip()) and not stripped.startswith("#")
+        ]
+
+
+def get_excludes(tracking_worktree_path: str, days: int, max_failures: int):
+    """
+    if a crate has reached `max_failures` failures in the last `days` days then we exclude it
+    """
+    failures = Counter()
+
+    failure_filenames = list(Path(tracking_worktree_path).glob("*-*-*"))
+    failure_filenames.sort(reverse=True)
+
+    for failure_filename in failure_filenames[:days]:
+        with open(failure_filename, "r") as file:
+            failures.update(
+                Counter(stripped for line in file if (stripped := line.strip()))
+            )
+
+    return [crate for crate, count in failures.items() if count >= max_failures]
+
+
+def get_repo_url() -> str:
+    repo_env_var = os.environ.get("GITHUB_REPOSITORY")
+    if repo_env_var:
+        return f"https://github.com/{repo_env_var}"
+    return subprocess.run(
+        ["gh", "repo", "view", "--json", "url", "--jq", ".url"],
+        capture_output=True,
+        text=True,
+    ).stdout.strip()
+
+
+def get_branch() -> str:
+    return subprocess.run(
+        ["git", "branch", "--show-current"],
+        capture_output=True,
+        text=True,
+    ).stdout.strip()
+
+
+def get_current_version_if_unbuilt(
+    repo_url: str,
+    crate: str,
+    target_arch: str,
+) -> CrateVersionDict | None:
+    version = get_latest_version(crate)
+    if not version:
+        return None
+
+    url = f"{repo_url}/releases/download/{crate}-{version['vers']}/{crate}-{version['vers']}-{target_arch}.tar.gz"
```

Comment:
I think using a python library for it would be more readable https://pypi.org/project/PyGithub/

For the rate-limit problem, I think we can provide a read-only GitHub token via secrets.

While `secrets.GITHUB_TOKEN` would work, it
 - has a rate limit of 1000 requests per hour
 - has too many permissions within the repo

If we create a read-only token using a robot account (which I've setup), we will have:
 - 5000 requests per hour
 - has read-only permission and cannot access any private info

Reply:
@alsuren if you think this is too much work for this PR, we could leave this to a follow-up PR.

Reply:
My main goals in this PR are to stop using the old stats server and to delete 400 lines of bash, so I am going to leave it like this.

I also quite like the symmetry of using exactly the same url that the quickinstall client uses. The old code even used the same http client (`curl`) but I decided that the connection pooling from requests was too much of a win.

> We could use the github api to get all artifacts and cache it.

I'm a bit wary of using a "do a lot of work and rely on caches to save us" approach. I've so far been taking the approach of "do as little work as possible and still make forward progress". I realise that there is a cutover though. As we check more crates every hour, we will want to have some kind of cached pre-filter to reduce the number of http requests we need to make. We would need to be wary of stale caches as well: if the cache says the package is there then we can trust that, but if the cache says it's not there, we need to re-check to make sure we didn't build it in the last hour.

In another unrelated note: I think it would be really interesting to gather stats around our build processes as well as our client requests. This would allow us to make better decisions when we make tweaks to the scheduling algorithm. I'm quite annoyed that the new influxdb cloud org doesn't have good notebook/dashboards. I really don't want to have to plug it into grafana just to share some graphs.